nhrpd/nhrp_cache.c

```diff
@@ -119,12 +119,50 @@ static void nhrp_cache_update_route(struct nhrp_cache *c)
 {
 	struct prefix pfx;
 	struct nhrp_peer *p = c->cur.peer;
+	char buf[3][SU_ADDRSTRLEN];
+	struct nhrp_interface *nifp;
 
 	sockunion2hostprefix(&c->remote_addr, &pfx);
 
 	if (p && nhrp_peer_check(p, 1)) {
-		netlink_update_binding(p->ifp, &c->remote_addr,
-				       &p->vc->remote.nbma);
+		if (sockunion_family(&c->cur.remote_nbma_natoa) != AF_UNSPEC) {
+			/* remote nbma is already set. Therefore, this is a
+			 * shortcut entry. Peer should NOT be changed to the
+			 * vc's remote nbma.
+			 */
+			debugf(NHRP_DEBUG_KERNEL,
+			       "cache (peer check ok, remote nbma set): "
+			       "Update binding for %s dev %s from (deleted) "
+			       "peer.vc.nbma %s to %s",
+			       sockunion2str(&c->remote_addr, buf[0],
+					     sizeof(buf[0])),
+			       p->ifp->name,
+			       sockunion2str(&p->vc->remote.nbma, buf[1],
+					     sizeof(buf[1])),
+			       sockunion2str(&c->cur.remote_nbma_natoa, buf[2],
+					     sizeof(buf[2])));
+
+			netlink_update_binding(p->ifp, &c->remote_addr,
+					       &c->cur.remote_nbma_natoa);
+		} else {
+			// update remote nbma to peer->vc->remote->nbma
+			debugf(NHRP_DEBUG_KERNEL,
+			       "cache (peer check ok, no remote nbma): "
+			       "Update binding for %s dev %s from (deleted) "
+			       "to peer.vc.nbma %s because "
+			       "c->cur.remote_nbma_natoa is %s",
+			       sockunion2str(&c->remote_addr, buf[0],
```

Comment:
if nbma natoa is unspec, no need to display it.
the log message should be related to nat. right?

Reply:
fixed